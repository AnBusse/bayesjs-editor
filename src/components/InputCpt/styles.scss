.input-cpt {
  min-width: 100px;
  border: none;
  outline: none;
  background-color: transparent;
}
