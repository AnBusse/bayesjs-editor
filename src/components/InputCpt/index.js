import React from 'react';
import float from 'float';
import { clamp, path, pipe } from 'ramda';
import PropTypes from 'prop-types';
import { noop } from 'lodash';
import { NODE_CPT_PRECISION } from 'constants/node';
import styles from './styles.scss';

const clamp0And1 = clamp(0, 1);
const pathTargetValue = path(['target', 'value']);

const formatValue = value => float.round(clamp0And1(value), NODE_CPT_PRECISION);
const getValueAndFormat = pipe(pathTargetValue, formatValue);

const onChangeHandler = onChange => event => onChange(event, getValueAndFormat(event));

const InputCpt = ({ onChange, ...props }) => (
  <input
    className={styles.inputCpt}
    type="number"
    step="0.01"
    max="1"
    min="0"
    onChange={onChangeHandler(onChange)}
    {...props}
  />
);

InputCpt.propTypes = {
  onChange: PropTypes.func,
};

InputCpt.defaultProps = {
  onChange: noop,
};

export default InputCpt;
