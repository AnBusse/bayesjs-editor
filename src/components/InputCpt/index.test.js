import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import InputCpt from './index';

const renderComponent = (props) => {
  const { container } = render(
    <InputCpt {...props} />,
  );

  return container.firstChild;
};

describe('InputCpt Component', () => {
  let component;
  const defaultProps = {
    id: 'id',
    onChange: jest.fn(),
    value: 0,
  };

  beforeEach(() => {
    defaultProps.onChange.mockClear();
    component = renderComponent(defaultProps);
  });

  it('matches snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  describe('When the value type is higher than one', () => {
    it('returns one as new value', () => {
      fireEvent.change(component, { target: { value: 2 } });

      expect(defaultProps.onChange).toHaveBeenCalledWith(1);
    });
  });

  describe('When the value type is between zero and one', () => {
    it('returns the value', () => {
      fireEvent.change(component, { target: { value: 0.5 } });

      expect(defaultProps.onChange).toHaveBeenCalledWith(0.5);
    });
  });

  describe('When the value type is lower than zero', () => {
    it('returns zero as new value', () => {
      fireEvent.change(component, { target: { value: -1 } });

      expect(defaultProps.onChange).toHaveBeenCalledWith(0);
    });
  });
});
