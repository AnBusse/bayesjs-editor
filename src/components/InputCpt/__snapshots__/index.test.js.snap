// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputCpt Component matches snapshot 1`] = `
<input
  class="inputCpt"
  data-testid="input-cpt__id"
  max="1"
  min="0"
  step="0.01"
  type="number"
  value="0"
/>
`;
